### PR TITLE
Increase the TensorFlow image version and fix a few typos.

### PR DESCRIPTION
**Pull request by ruddyscent**
Many students are failing to train the models for the 'object detection in urban environment' project. We can find [advice](https://knowledge.udacity.com/questions/984202#984205) on the Knowlege platform to use an instance of ml.m4.2xlarge and reduce `num_train_steps` to 200, but the trained model with such few steps fails at object detection.

If I raise the version of the TensorFlow image to **2.12.0-gpu**, I could successfully train on the following models on **ml.p3.2xlarge**(with GPU).
* SSD MobileNet V2 FPNLite 640x640
* SSD ResNet50 V1 FPN 640x640 (RetinaNet50)
* EfficientDet D1 640x640

However, training failed for the following models, and I still need to perform a detailed analysis of the reasons for the failure.
* Faster R-CNN ResNet50 V1 640x640
* Faster R-CNN ResNet152 V1 640x640

### CODE REVIEW

**Comment by islamalsawy** [comment]
@ruddyscent Thanks for the feedback. Do you have any updates regarding the other two models? So that you can start fixing it as a task.

**Comment by ruddyscent** [comment]
I ran a test using tensorflow 2.14.0, and here are the results.

I could train the following models:
* EfficientDet D1 640x640
* SSD MobileNet V2 FPNLite 640x640
* Faster R-CNN ResNet50 V1 640x640

I could not test the following model yet:
* SSD MobileNet V1 FPN 640x640 

The following model is being trained and gets an error about mismatched layer shapes:
* Faster R-CNN ResNet152 V1 640x640

**Comment by ruddyscent** [comment]
Using TF 2.13, **training of all models is possible**. With TF 2.14, I failed to train all models.

As you can see in case 6, it takes too long to train without GPU.


All cases use ~35% or less of memory over the whole process. We don't have to use the `2xlarge` or higher instance; we can guide students to reduce the batch size if memory usage is an issue.

`ml.g5.xlarge` is fast, not only in training but also in training preparation. Regarding GPU performance, I wouldn't recommend anything less than an A10G, which is what `ml.g5.xlarge` uses, as it would make it difficult for students to find appropriate hyperparameters through trial and error.

The following are the training times measured with a few instance configurations:
num_train_steps: "2000"
sagemaker_region: "us-east-1"
|case|model|instance|batch size|preparation (min)|training (min)|
|--|--|--|--|--|--|
|1|EfficientDet D1 640x640|ml.g4dn.xlarge|8|20|60|
|2|EfficientDet D1 640x640|ml.g5.xlarge|8|5|32|
|3|SSD MobileNet V2 FPNLite 640x640|ml.g5.xlarge|8|5|15|
|4|SSD ResNet50 V1 FPN 640x640 (RetinaNet50)|ml.g5.xlarge|8|5|21|
|5|Faster R-CNN ResNet50 V1 640x640|ml.g5.xlarge|2|5|11|
|6|Faster R-CNN ResNet50 V1 640x640|ml.m4.xlarge|2|12|>180|
|7|Faster R-CNN ResNet152 V1 640x640|ml.g5.xlarge|2|5|15|

**Comment by hanneskannes123** [comment]
@ruddyscent please note that the Udacity-provided AWS account does not have any quota for `ml.g5.xlarge`

> ResourceLimitExceeded: An error occurred (ResourceLimitExceeded) when calling the CreateTrainingJob operation: The account-level service limit 'ml.g5.xlarge for training job usage' is 0 Instances, with current utilization of 0 Instances and a request delta of 1 Instances. Please use AWS Service Quotas to request an increase for this quota. If AWS Service Quotas is not available, contact AWS support to request an increase for this quota.

**Comment by ruddyscent** [comment]
@hanneskannes123 It's odd. I could make a `ml.g5.xlarge` instance through Udacity Cloud Gateway. Would you try in `us-east-1` region?

**Comment by hanneskannes123** [comment]
@ruddyscent I double checked that I am in `us-east-1` and still get the same error for `ml.g5.xlarge`. `ml.p3.2xlarge` and `ml.m5.2xlarge` are working.

Edit: Attached a screenshot of the service quotas
Screenshot 2023-10-26 at 14 52 43

**Comment by islamalsawy** [comment]
@hanneskannes123 Kindly, did you try the cloud gateway from inside the Self-driving ND or another ND? Could you share your email to verify?

**Comment by hanneskannes123** [comment]
> @hanneskannes123 Kindly, did you try the cloud gateway from inside the Self-driving ND or another ND? Could you share your email to verify?

@islamalsawy I shared my email address via LinkedIn DM as I don't want to post it here in public.

**Comment by islamalsawy** [comment]
@hanneskannes123 Thanks so much for letting me know that you were able to solve the quota error by requesting quota increase.

File: 2_run_inference/2_deploy_model.ipynb [hunk 1]
```diff
@@ -48,16 +49,17 @@
...
   {
+   "attachments": {},
    "cell_type": "markdown",
    "id": "c03b56d7",
    "metadata": {},
    "source": [
     "Now that you have trained successfully your model, you want to look at the predictions on some sample images. To do so, you will need to find the s3 path of the exported model. You can navigate to the Training jobs section of the AWS web UI and click on the training job of interest. Scroll down and you should see something like this:\n",
     "\n",
-    "![Example Artefact](../data/example_artefact.png)\n",
+    "![Example Artifact](../data/example_artefact.png)\n",
```
Comment (islamalsawy) [inline 2_run_inference/2_deploy_model.ipynb]: @ruddyscent Could you please change the file name to match your edit? example_**artifact**.png

File: 1_model_training/docker/Dockerfile [hunk 2]
```diff
@@ -21,7 +14,7 @@ RUN apt-get update && apt-get install -y \
     libgl1-mesa-dev \
     wget
 
-# Copy this version of of the model garden into the image
+# Copy this version of the model garden into the image
 COPY models/research/object_detection /home/tensorflow/models/research/object_detection
```
Comment (islamalsawy) [inline 1_model_training/docker/Dockerfile:19]: @ruddyscent Kindly, it seems there is an attribute error that appears in logs which does not affect training completion or trained model's deployment. By making a quick research, we can find this error is related to the version of `pillow` which could be resolved by using:


`!pip install pillow==9.5`
or
`!pip install pillow --upgrade`

However, I think it does not work by installing manually, so do you have any workarounds here, please?

image

File: 2_run_inference/2_deploy_model.ipynb (hunk 1 above)
Comment (ruddyscent) [reply 2_run_inference/2_deploy_model.ipynb]: I've incorporated your suggestion.

File: 1_model_training/docker/Dockerfile (hunk 2 above)
Comment (ruddyscent) [reply 1_model_training/docker/Dockerfile:19]: We need to change the pillow version of the image that runs in SageMaker, not the environment that runs 1_train_model.ipynb. I've tested with the version you provided, and it resolved the problem. So, I committed the fix.

**Pull request merged**